scipy/stats/stats.py

```diff
@@ -408,7 +411,21 @@ def gmean(a, axis=0, dtype=None):
             log_a = np.log(np.asarray(a, dtype=dtype))
     else:
         log_a = np.log(a)
-    return np.exp(log_a.mean(axis=axis))
+
+    if weights:
+        if not isinstance(weights, np.ndarray):
+            # if not an ndarray object attempt to convert it
+            weights = np.array(weights, dtype=dtype)
+        elif dtype:
+            # Must change the default dtype allowing array type
+            if isinstance(a, np.ma.MaskedArray):
+                weights = np.ma.asarray(weights, dtype=dtype)
+            else:
+                weights = np.asarray(weights, dtype=dtype)
+
+        assert weights.shape == log_a.shape, "Shape of a and weights must be identical."
```

Comment:
Please raise an appropriate error type.

Reply:
Ok. Would a `ValueError` be appropriate in your opinion?

Reply:
Yes, I think so.

Reply:
Resolved in 6a0d8258ddd896ebe35babc9ea583e874a139115.